### PR TITLE
feat: create first version of skeleton for kakarot ssj

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [x] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/eoa/eth_transaction.cairo [hunk 1]
```diff
@@ -0,0 +1,61 @@
+use array::ArrayTrait;
+
+struct EthereumTransaction {
+    nonce: felt252,
```
[the remark this answers is not in the document]
Comment (Eikix) [reply src/eoa/eth_transaction.cairo:4]: unsure, open to challenge, what should we use here? u128?

File: src/eoa/eth_transaction.cairo [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+trait EthTransaction {
+    /// Decode a legacy Ethereum transaction
+    /// This function decodes a legacy Ethereum transaction in accordance with EIP-155.
+    /// It returns transaction details including nonce, gas price, gas limit, destination address, amount, payload,
+    /// transaction hash, and signature (v, r, s). The transaction hash is computed by keccak hashing the signed
+    /// transaction data, which includes the chain ID in accordance with EIP-155.
+    /// # Arguments
+    /// tx_data The raw transaction data
+    fn decode_legacy_tx(tx_data: @Array<u8>) -> EthereumTransaction;
```
[the remark this answers is not in the document]
Comment (Eikix) [reply src/eoa/eth_transaction.cairo:24]: i see

File: src/eoa/eth_transaction.cairo [hunk 3]
```diff
@@ -0,0 +1,61 @@
+use array::ArrayTrait;
+
+struct EthereumTransaction {
+    nonce: felt252,
+    gas_price: felt252,
+    gas_limit: felt252,
+    destination: felt252,
+    amount: felt252,
+    payload: @Array<felt252>,
```
[the remark this answers is not in the document]
Comment (Eikix) [reply src/eoa/eth_transaction.cairo:9]: alright

File: src/eoa/eth_transaction.cairo (hunk 1 above)
[the remark this answers is not in the document]
Comment (Eikix) [reply src/eoa/eth_transaction.cairo:4]: i'd like to use a big value, but u256 takes up two felts

[the remark this answers is not in the document]
Comment (enitrat) [reply src/eoa/eth_transaction.cairo:4]: If it's not stored on chain I would say that it doesn't matter a lot? Ideally we would have a `u250` native type which would solve this problem

File: src/utils/helpers.cairo [hunk 4]
```diff
@@ -240,3 +240,34 @@ impl U256TryIntoU128 of TryInto<u256, u128> {
...
+        return b;
+    }
+}
+
+// Raise a number to a power.
+/// * `base` - The number to raise.
+/// * `exp` - The exponent.
+/// # Returns
+/// * `felt252` - The result of base raised to the power of exp.
+fn pow(base: felt252, exp: felt252) -> felt252 {
```
Comment (ClementWalter) [inline src/utils/helpers.cairo:258]: can you create an issue to use fast exponentiation

File: src/eoa/eth_transaction.cairo (hunk 1 above)
[the remark this answers is not in the document]
Comment (ClementWalter) [reply src/eoa/eth_transaction.cairo:4]: isn't `felt252` the native type?

[the remark this answers is not in the document]
Comment (enitrat) [reply src/eoa/eth_transaction.cairo:4]: we should work with unsigned integers as much as possible. After discussing this with Ori, he advised to use either `u128` or `u256` depending on use case (here, if u128 is not enough we can use u256). 

Optimisations principally matter when you 're storing something on chain

File: src/utils/helpers.cairo [hunk 5]
```diff
@@ -240,3 +240,34 @@ impl U256TryIntoU128 of TryInto<u256, u128> {
...
+        return 1;
+    } else {
+        return base * pow(base, exp - 1);
+    }
+}
+
+/// Panic with a custom code.
+/// # Arguments
+/// * `code` - The code to panic with. Must be a short string to fit in a felt252.
+fn panic_with_code(code: felt252) {
```
Comment (enitrat) [inline src/utils/helpers.cairo:269]: This is no longer required, now we have `panic_with_felt252`

File: src/instructions.cairo [hunk 6]
```diff
@@ -69,7 +69,7 @@ impl EVMInstructionsImpl of EVMInstructionsTrait {
 
         // Check if PC is not out of bounds.
         if pc >= bytecode_len {
-            utils::panic_with_code(0);
+            helpers::panic_with_code(0);
```
Comment (enitrat) [inline src/instructions.cairo:72]: `panic_with_felt252(0);`

File: src/eoa/eth_transaction.cairo (hunk 1 above)
[the remark this answers is not in the document]
Comment (Eikix) [reply src/eoa/eth_transaction.cairo:4]: nice call
let's reuse reth's types
https://github.com/paradigmxyz/reth/blob/500b0fac135fe07635d871d64467326599e2b27e/crates/primitives/src/transaction/mod.rs#L162

[the remark this answers is not in the document]
Comment (Eikix) [reply src/eoa/eth_transaction.cairo:4]: my bad -> we mostly use those types in kakarot-rpc https://github.com/paradigmxyz/reth/blob/500b0fac135fe07635d871d64467326599e2b27e/crates/rpc/rpc-types/src/eth/transaction/mod.rs

not the one above

[the remark this answers is not in the document]
Comment (Eikix) [reply src/eoa/eth_transaction.cairo:4]: will ask on reth tg chat

File: src/utils/helpers.cairo (hunk 4 above)
Comment (Eikix) [reply src/utils/helpers.cairo:258]: ok! #65

File: src/utils/helpers.cairo (hunk 5 above)
Comment (Eikix) [reply src/utils/helpers.cairo:269]: ok!

File: src/instructions.cairo (hunk 6 above)
Comment (Eikix) [reply src/instructions.cairo:72]: okl!

File: src/eoa/eth_transaction.cairo (hunk 1 above)
[the remark this answers is not in the document]
Comment (Eikix) [reply src/eoa/eth_transaction.cairo]: done

**Review by greged93 — APPROVED** [review]
lgtm

**Pull request merged**